pkcs11/pkcs11.go

```diff
@@ -1687,3 +1706,169 @@ func (s *Slot) generateECDSA(o keyOptions) (crypto.PrivateKey, error) {
+	rv = C.ck_encrypt(r.o.fl, r.o.h, &cDataBytes[0], C.CK_ULONG(len(cDataBytes)), nil, &cCipherLen)
+	if err := isOk("C_Encrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cCipher := make([]C.CK_BYTE, cCipherLen)
+
+	rv = C.ck_encrypt(r.o.fl, r.o.h, &cDataBytes[0], C.CK_ULONG(len(cDataBytes)), &cCipher[0], &cCipherLen)
+	if err := isOk("C_Encrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cipher := toBytes(cCipher)
+
+	return cipher, nil
+}
+
+func (r *rsaPrivateKey) decryptRSA(encryptedData []byte) ([]byte, error) {
+	hash := r.getHash()
+	cParam, err := makeCParamRSAOAEP(hash)
+	if err != nil {
+		return nil, err
+	}
+	defer C.free(unsafe.Pointer(cParam))
+
+	cEncDataBytes := toCBytes(encryptedData)
+
+	m := makeMechanism(C.CKM_RSA_PKCS_OAEP, C.CK_VOID_PTR(cParam), C.sizeof_CK_RSA_PKCS_OAEP_PARAMS)
+
+	rv := C.ck_decrypt_init(r.o.fl, r.o.h, &m, r.o.o)
+	if err := isOk("C_DecryptInit", rv); err != nil {
+		return nil, err
+	}
+
+	var cDecryptedLen C.CK_ULONG
+
+	// First call is used to determine length necessary to hold decrypted data
+	rv = C.ck_decrypt(r.o.fl, r.o.h, &cEncDataBytes[0], C.CK_ULONG(len(cEncDataBytes)), nil, &cDecryptedLen)
+	if err := isOk("C_Decrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cDecrypted := make([]C.CK_BYTE, cDecryptedLen)
+
+	rv = C.ck_decrypt(r.o.fl, r.o.h, &cEncDataBytes[0], C.CK_ULONG(len(cEncDataBytes)), &cDecrypted[0], &cDecryptedLen)
+	if err := isOk("C_Decrypt", rv); err != nil {
+		return nil, err
+	}
+
+	decrypted := toBytes(cDecrypted)
+
+	return decrypted, nil
+}
+
+func toBytes(data []C.CK_BYTE) []byte {
+	goBytes := make([]byte, len(data))
+	for i, b := range data {
+		goBytes[i] = byte(b)
+	}
+	return goBytes
+}
+
+func toCBytes(data []byte) []C.CK_BYTE {
+	cBytes := make([]C.CK_BYTE, len(data))
+	for i, b := range data {
+		cBytes[i] = C.CK_BYTE(b)
+	}
+	return cBytes
+}
+
+func makeMechanism(mech C.CK_MECHANISM_TYPE, paramPtr C.CK_VOID_PTR, paramLen int) C.CK_MECHANISM {
+	mechanism := C.CK_MECHANISM{
+		mechanism:      mech,
+		pParameter:     paramPtr,
+		ulParameterLen: C.CK_ULONG(paramLen),
+	}
+	return mechanism
+}
+
+func makeCParamRSAOAEP(hash *crypto.Hash) (C.CK_RSA_PKCS_OAEP_PARAMS_PTR, error) {
+	cParam := (C.CK_RSA_PKCS_OAEP_PARAMS_PTR)(C.malloc(C.sizeof_CK_RSA_PKCS_OAEP_PARAMS))
+
+	switch *hash {
+	case crypto.SHA256:
+		cParam.hashAlg = C.CKM_SHA256
+		cParam.mgf = C.CKG_MGF1_SHA256
+	case crypto.SHA384:
+		cParam.hashAlg = C.CKM_SHA384
+		cParam.mgf = C.CKG_MGF1_SHA384
+	case crypto.SHA512:
+		cParam.hashAlg = C.CKM_SHA512
+		cParam.mgf = C.CKG_MGF1_SHA512
+	case crypto.SHA1:
+		cParam.hashAlg = C.CKM_SHA_1
+		cParam.mgf = C.CKG_MGF1_SHA1
+	default:
+		return nil, fmt.Errorf("makeCParamRSAOAEP error, unsupported hash algorithm: %s", hash)
+	}
+
+	cParam.source = C.CKZ_DATA_SPECIFIED
+	cParam.pSourceData = nil
+	cParam.ulSourceDataLen = 0
+
+	return cParam, nil
+}
+
+func Decrypt(key crypto.PrivateKey, encryptedData []byte) ([]byte, error) {
```

Comment:
remove this in favor of users type casting to https://pkg.go.dev/crypto#Decrypter

For example, see how the TLS package uses this:

https://github.com/golang/go/blob/2efebf3ee0796054942a49d0f8b1c2c50193c7fe/src/crypto/tls/key_agreement.go#L59